### ISSUE TITLE
grafana-server: crash on concurrent map write

### ISSUE THREAD

**Issue by Robertof** [grafana/grafana#67960]
### 
# What went wrong?

Grafana panicked during start up on my system with a concurrent map write inside `gopkg.in/ini.v1`. I believe this might be due to concurrent requests coming in as Grafana was starting up (and still hadn't finished processing the config file, perhaps).

[Full log](https://github.com/grafana/grafana/files/11413027/Explore-logs-2023-05-06.18.33.13.txt)

Relevant excerpt:

```
2023-05-02T12:11:40+01:00	fatal error: concurrent map writes
2023-05-02T12:11:40+01:00	goroutine 13017 [running]:
2023-05-02T12:11:40+01:00	gopkg.in/ini%2ev1.(*File).NewSection(0x4000d25200, {0x412db1d?, 0x40015f1101?})
2023-05-02T12:11:40+01:00		gopkg.in/ini.v1@v1.67.0/file.go:180 +0x78
2023-05-02T12:11:40+01:00	gopkg.in/ini%2ev1.(*File).Section(0x38e78c0?, {0x412db1d, 0x9})
2023-05-02T12:11:40+01:00		gopkg.in/ini.v1@v1.67.0/file.go:108 +0x3f4
2023-05-02T12:11:40+01:00		github.com/grafana/grafana/pkg/setting/setting.go:1304
2023-05-02T12:11:40+01:00	github.com/grafana/grafana/pkg/setting.(*Cfg).SectionWithEnvOverrides(...)
[...]
2023-05-02T12:11:40+01:00	goroutine 13026 [running]:
[...]
2023-05-02T12:11:40+01:00	net/http.HandlerFunc.ServeHTTP(0x4001b359c7?, {0x4a70e20?, 0x4002f86d20?}, 0x40?)
2023-05-02T12:11:40+01:00		github.com/grafana/grafana/pkg/api/response/web_hack.go:40 +0x38
2023-05-02T12:11:40+01:00	github.com/grafana/grafana/pkg/api/response.wrap_handler.func3({0x4a70e20?, 0x4002f86d20?}, 0x4001483f38?)
2023-05-02T12:11:40+01:00		github.com/grafana/grafana/pkg/api/index.go:170 +0x24
2023-05-02T12:11:40+01:00	github.com/grafana/grafana/pkg/api.(*HTTPServer).Index(0x40011f2000, 0x4000240770)
2023-05-02T12:11:40+01:00		github.com/grafana/grafana/pkg/api/index.go:38 +0x1a8
2023-05-02T12:11:40+01:00	github.com/grafana/grafana/pkg/api.(*HTTPServer).setIndexViewData(0x40011f2000, 0x4000240770)
2023-05-02T12:11:40+01:00		github.com/grafana/grafana/pkg/api/frontendsettings.go:149 +0x1b24
2023-05-02T12:11:40+01:00	github.com/grafana/grafana/pkg/api.(*HTTPServer).getFrontendSettingsMap(0x40011f2000, 0x4000240770)
2023-05-02T12:11:40+01:00		github.com/grafana/grafana/pkg/setting/setting.go:1304
2023-05-02T12:11:40+01:00	github.com/grafana/grafana/pkg/setting.(*Cfg).SectionWithEnvOverrides(...)
2023-05-02T12:11:40+01:00		gopkg.in/ini.v1@v1.67.0/file.go:180 +0x78
2023-05-02T12:11:40+01:00	gopkg.in/ini%2ev1.(*File).Section(0x38e78c0?, {0x412db1d, 0x9})
2023-05-02T12:11:40+01:00		gopkg.in/ini.v1@v1.67.0/file.go:108 +0x3f4
2023-05-02T12:11:40+01:00	gopkg.in/ini%2ev1.(*File).NewSection(0x4000d25200, {0x412db1d?, 0x400177f101?})
```

At a glance, this looks like a race in the usage of the `ini` package. `NewSection` modifies `(*File).sections`: https://github.com/go-ini/ini/blob/842b9a946dd3bb45b95001ce6d9ae86f3b3ce921/file.go#L82-L111

Grafana calls `.Section()` (which invokes `NewSection` if the section isn't already found) without locking: https://github.com/grafana/grafana/blob/15d41698138f250b85ccd423f2f252dea7cf3c93/pkg/setting/setting.go#L1327-L1329

It seems like `go-ini` has a `BlockMode` flag that enables locking on those methods, but I have no idea if Grafana uses it. 

### 
# How do we reproduce it?

Unfortunately, I don't have a reproducer. This crash happened only once after a reboot of my system. The only reproduction steps I can think of is just of multiple requests coming in during bootstrap (in my case at an embedded dashboard). As such, this might be able to be emulated by just hammering a valid `/d-solo/` endpoint during startup.

### 
# What Grafana version are you using?

v9.4.3

### 
### Grafana Platform?

Other

### User's OS?

NixOS

### Is this a Regression?

None

**Comment by tonypowa** [comment]
Hi @Robertof 

Thank you for reporting this issue

This error has been seen before 
https://github.com/grafana/grafana/issues/44509

I'd advise you to add a comment in the thread above to leave a record of this, specially because of not being reproducible.

/duplicate of #44509

**Comment by Robertof** [comment]
Thanks for triaging this @tonypowa! I don't necessarily agree that it's the same issue -- the file and cause are the same (race in `go-ini` related code), but the crash in #44509 happens here:

https://github.com/grafana/grafana/blob/15d41698138f250b85ccd423f2f252dea7cf3c93/pkg/setting/setting.go#L1310-L1323

whilst mine happens here:

https://github.com/grafana/grafana/blob/15d41698138f250b85ccd423f2f252dea7cf3c93/pkg/setting/setting.go#L1327-L1329

I'll leave a comment there regardless 👍

**Comment by tonypowa** [comment]
thank you @Robertof 

I would have sent this issue to the same team, the back-end team, anyway so they should be able to assist you further from that ticket
